src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/SpatioTemporalZonalStatistics.java

```diff
@@ -0,0 +1,583 @@
+                    CoordinateReferenceSystem dataCrs = reader.getCoordinateReferenceSystem();
+                    CoordinateReferenceSystem zonesCrs =
+                            builder.getFeatureType().getGeometryDescriptor().getCoordinateReferenceSystem();
+                    if (!CRS.equalsIgnoreMetadata(zonesCrs, dataCrs)) {
+                        zoneGeom = JTS.transform(zoneGeom, CRS.findMathTransform(zonesCrs, dataCrs, true));
+                    }
+
+                    // gather the statistics
+                    StatisticsAggregator stats = processStatistics(zoneGeom);
+
+                    // build the resulting feature
+                    if (stats != null) {
+                        builder.addAll(zone.getAttributes());
+                        addStatsToFeature(stats, requestedStats);
+                        features.add(builder.buildFeature(zone.getID()));
+
+                    } else {
+                        builder.addAll(zone.getAttributes());
+                        features.add(builder.buildFeature(zone.getID()));
+                    }
+                } catch (Exception e) {
+                    throw new ProcessException("Failed to compute statistics on feature " + zone, e);
+                }
+            }
+            // return the first feature in the current buffer
+            SimpleFeature f = features.remove(0);
+            return f;
+        }
+
+        /** Add the statistics to the feature builder */
+        private void addStatsToFeature(StatisticsAggregator stats, Set<Statistic> requestedStats) {
+            double count = stats.getAggregatedCount();
+            builder.add(count); // count
+            addDynamicStatsToFeature(builder, stats, requestedStats);
+        }
+
+        private void addDynamicStatsToFeature(
+                SimpleFeatureBuilder builder, StatisticsAggregator stats, Set<Statistic> requestedStats) {
+            if (requestedStats.contains(Statistic.MIN)) {
+                builder.add(stats.getAggregatedMin());
+            }
+            if (requestedStats.contains(Statistic.MAX)) {
+                builder.add(stats.getAggregatedMax());
+            }
+            if (requestedStats.contains(Statistic.SUM)) {
+                builder.add(stats.getAggregatedSum());
+            }
+            if (requestedStats.contains(Statistic.MEAN)) {
+                builder.add(stats.getAggregatedMean());
+            }
+            if (requestedStats.contains(Statistic.MEDIAN)) {
+                builder.add(stats.getAggregatedMedian());
+            }
+        }
+
+        private StatisticsAggregator processStatistics(Geometry geometry) throws TransformException, IOException {
+            GridCoverage2D refCoverage = null;
+            GridCoverage2D dataCoverage = null;
+            GridCoverage2D cropped = null;
+            ReferencedEnvelope geometryEnvelope = null;
+            List<Range<Double>> noDataValueRangeList = null;
+            ROI roi = null;
```

Comment:
This is not a C program, are you sure all these variables cannot be declared closer to their first usage point? (e..g, the coverages could be declared just before the try for example)

Reply:
Right. beside the 2 coverages, all the other variables are initialized once and reused within the for loop.
I'm moving the 2 coverages closer to their usage.